src/workload.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""MySQL Router workload"""
+
+import dataclasses
+import logging
+import pathlib
+import string
+import typing
+
+import ops
+
+import mysql_shell
+
+if typing.TYPE_CHECKING:
+    import charm
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(kw_only=True)
+class Workload:
+    """MySQL Router workload"""
+
+    _container: ops.Container
+
+    CONTAINER_NAME = "mysql-router"
+    _SERVICE_NAME = "mysql_router"
+
+    @property
+    def container_ready(self) -> bool:
+        """Whether container is ready"""
+        return self._container.can_connect()
+
+    @property
+    def _enabled(self) -> bool:
+        """Service status"""
+        service = self._container.get_services(self._SERVICE_NAME).get(self._SERVICE_NAME)
+        if service is None:
+            return False
+        return service.startup == "enabled"
+
+    @property
+    def version(self) -> str:
+        """MySQL Router version"""
+        process = self._container.exec(["mysqlrouter", "--version"])
+        raw_version, _ = process.wait_output()
+        for version in raw_version.split():
+            if version.startswith("8"):
+                return version
+        return ""
+
+
+@dataclasses.dataclass(kw_only=True)
+class AuthenticatedWorkload(Workload):
+    """Workload with connection to MySQL cluster"""
+
+    _admin_username: str
+    _admin_password: str
+    _host: str
+    _port: str
+    _charm: "charm.MySQLRouterOperatorCharm"
+
+    _UNIX_USERNAME = "mysql"
+    _ROUTER_CONFIG_DIRECTORY = pathlib.Path("/etc/mysqlrouter")
+    _ROUTER_CONFIG_FILE = "mysqlrouter.conf"
+    _TLS_CONFIG_FILE = "tls.conf"
+    _TLS_KEY_FILE = "custom-key.pem"
+    _TLS_CERTIFICATE_FILE = "custom-certificate.pem"
+
+    @property
+    def _router_username(self) -> str:
+        unit_id = self._charm.unit.name.split("/")[1]
+        return f"mysqlrouter_{unit_id}"
```

Comment:
thoughts on including the name of the application (which is unique in a model) in the router username? this is because if 2 router applications are related to the same cluster, then there may be a conflict in usernames